### ISSUE TITLE
PHP 8.0.0 compatibility

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1546]
According to https://wiki.php.net/todo/php80 we should expect PHP 8.0.0 RC1 in October. GA is currently scheduled for November 26th. While I can see the benefit of making use of the new whizzy stuff that PHP 8 brings (see https://github.com/textpattern/textpattern/issues/1506#issuecomment-648379064 and https://github.com/textpattern/textpattern/issues/1506#issuecomment-648390434 for commentary from @Bloke ), the more pressing issue is to make sure Textpattern works with PHP 8.0.0 with as few known issues as possible.

I can (and will) get the PHP 8 demo server running so we can batter it around, and I have this lofty idea that we could essentially reserve the 4.8.4 release as the "works with PHP 8" version and release it on the same day as PHP 8.0.0 GA.

Yes, it's another multiple branch situation, which isn't ideal, but hopefully there isn't a vast amount of work to be accomplished in the relatively short time.

As a guide, from the time of this post, and assuming no movement on the dates listed:

* 20 days until RC1 is released (18 days until it's tagged and I grab it)
* 62 days until RC5 is released (60 days until it's tagged and I grab it)
* …and 76 days until GA is released (74 days if you're me, and so on)

If this has legs, it's straightforward enough to have `dev` and (gasp) `4.8.4` branches running atop PHP RCs pretty soon, so we can road test both in parallel. And if I have my act together week of GA, I can grab the code and we can know for certain that what will be PHP 8.0.0 two days (-ish) before the public release. Prep the release, job done.

That's what I'm thinking - what do you think?

**Comment by bloatware** [comment]
> reserve the 4.8.4 release as the "works with PHP 8" version and release it on the same day as PHP 8.0.0 GA

Yep, we must get php8 ready, and hopefully 4.8.3 has no urgent bugs to patch.

**Comment by bloatware** [comment]
Shouldn't we create a 4.8.4 branch already? It's easier to commit there and merge into dev than the opposite way.

**Comment by Bloke** [comment]
Sure, I guess it makes sense if we know we're going to release it to coincide with PHP 8.

Leave it with me, I'll make one now.

**Comment by bloatware** [comment]
We can safely delete it later if it is finally not needed, right?

**Comment by Bloke** [comment]
Yep. Easy to blat if we don't need it.

[4.8.4 branch created](https://github.com/textpattern/textpattern/commit/47926024f016286d14405a26a2a98065aaedd1be)

**Comment by petecooper** [comment]
For reference, RC1 was tagged 29 September and will be announced 1 October. I've installed it and we have two demos up: `dev` and `4.8.4`.

Assuming the PHP release schedule is adhered to, we should expect (timestamps relative to this post):

* RC2 in 14 days
* RC3 in 28 days
* RC4 in 42 days
* GA in 56 days

**Comment by petecooper** [comment]
For reference, expecting RC2 today.

**Comment by petecooper** [comment]
RC2 tagged 13 October and will be announced 15 October. Now installed, please see https://github.com/textpattern/textpattern/issues/1546#issuecomment-701323168 for links to dev sites (no changes to URLs).

Assuming the PHP release schedule is adhered to, we should expect (timestamps relative to this post):

* RC3 in 14 days
* RC4 in 28 days
* GA in 42 days

**Comment by petecooper** [comment]
Some possible weirdness with PHP8.0.0RC2 and crypto, investigating. Please ignore errors that refer to .pem file not being loaded.

**Comment by petecooper** [comment]
RC3 tagged 27 October and will be announced 29 October. Now installing to demo server, please see https://github.com/textpattern/textpattern/issues/1546#issuecomment-701323168 for links to dev sites (no changes to URLs).

Assuming the PHP release schedule is adhered to, we should expect (timestamps relative to this post):

* RC4 in 14 days
* GA in 28 days

**Comment by petecooper** [comment]
For reference, expecting RC4 tomorrow.

**Comment by petecooper** [comment]
RC4 tagged 27 October and will be announced 29 October. Now installing to demo server, please see https://github.com/textpattern/textpattern/issues/1546#issuecomment-701323168 for links to dev sites (no changes to URLs).

Assuming the PHP release schedule is adhered to, we should expect (timestamps relative to this post):

* RC4 in 14 days
* GA in 28 days

**Comment by petecooper** [comment]
RC4 was tagged 10 November and will be announced 12 November. Now installing to demo server, please see https://github.com/textpattern/textpattern/issues/1546#issuecomment-701323168 for links to dev sites (no changes to URLs).

Assuming the PHP release schedule is adhered to, we should expect GA in 14 days. There are no further release candidates planned for PHP 8.0.0, so this is as close as we'll get to a gold master before the end of November.

**Comment by petecooper** [comment]
PHP 8.0.0 wiki page has been updated in the last few days, and RC5 has just landed. I'll get that installed today, likely in the next hour or so.

Assuming the PHP release schedule is adhered to, we should expect GA in 6 days. There are no further release candidates planned for PHP 8.0.0, so this is as close as we'll get to a gold master before the end of November.

**Comment by petecooper** [comment]
For reference, expecting GA tomorrow.

**Comment by petecooper** [comment]
8.0.0 GA was tagged 24 November and will be announced 26 November. Now installing to demo server, please see https://github.com/textpattern/textpattern/issues/1546#issuecomment-701323168 for links to dev sites (no changes to URLs).

Assuming `dev` and `4.8.4` run broadly OK, I'll close this tracking issue and we can mobilise testers for the final furlongs before 4.8.4 lands.

**Comment by petecooper** [comment]
Closing.

**Comment by philwareham** [comment]
Just installed latest MAMP with PHP 8 running, getting this in the diagnostics...

<img width="612" alt="Screenshot 2020-12-16 at 09 24 03" src="https://user-images.githubusercontent.com/413665/102329932-dcefd100-3f80-11eb-888f-07aaa2781a1e.png">

Is GD not bundled with PHP any more in v8?

**Comment by petecooper** [comment]
>Is GD not bundled with PHP any more in v8?

![Screenshot 2020-12-16 at 09 30 06](https://user-images.githubusercontent.com/226578/102330463-75865100-3f81-11eb-961e-8f9f93efefe3.png)

It's compiled with `--enable-gd`, so definitely available. All depends on who compiles it as to whether it's available. The "bundled" term is a bit disingenuous, I think it refers to whether the GD library is the system-wide library or a specific one compiled in at build time. From https://www.php.net/manual/en/image.installation.php:

> To enable GD-support configure PHP --with-gd[=DIR], where DIR is the GD base install directory. To use the recommended bundled version of the GD library, use the configure option --with-gd. GD library requires libpng and libjpeg to compile. As of PHP 7.4.0, --with-gd becomes --enable-gd (whether to enable the extension at all) and --with-external-gd (to opt into using an external libgd, rather than the bundled one).

**Comment by philwareham** [comment]
Maybe an oversight by MAMP's devs? It is an initial release from them as of today.

Looks like you can't add ImageMagick automatically in MAMP Pro for PHP 8 at the moment (which you can for PHP 7.4.x in MAMP Pro).

<img width="451" alt="Screenshot 2020-12-16 at 09 40 14" src="https://user-images.githubusercontent.com/413665/102331519-db270d00-3f82-11eb-97c9-453e321c7252.png">

**Comment by petecooper** [comment]
FYI, compiling PHP with `--with-external-gd ` uses the operating system `libgd`, otherwise PHP uses the provided (bundled) version included with the PHP source.
